<!DOCTYPE html>
<html>
<head>
<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<title>Pan-Core / Comparative genomomics exercise</title>
<link rel="stylesheet" type="text/css" href="pancore.css">
<link href="http://fonts.googleapis.com/css?family=Headland+One|Life+Savers:400,700" rel="stylesheet" type="text/css">
</head>
<body style="text-align : justify; font-family: 'Life Savers', cursive; font-size: 24px">
<div>
<h3>Pan-Core / Comparative genomomics exercise <a href="https://docs.google.com/presentation/d/19kK3DAaT8pP6AkC2qrFoLwwc8CCaGQzZ6SSrfnlK-fc/edit?usp=sharing">(Slides)</a></h3>
<a href="mailto@sigmarkarl@gmail.com">sigmarkarl@gmail.com</a>
</div>
<div class="article">
<section>
<b>Data preparation</b>
<p>
Data was prepared as described in the <a href="/pancore.html">manual</a>. 
The preparation is a bit of a hassle so we have prepared few sets of microbial genomes for you.
Download one from here: <a href="http://130.208.181.105/allthermus.zip"><i>Thermus</i> species</a>, <a href="https://drive.google.com/file/d/0B7Hl6KFtpv4LOUFTMmNOTEtZSUU/edit?usp=sharing"><i>Spirochaeta</i> species</a>, <a href="https://drive.google.com/file/d/0B7Hl6KFtpv4LRm5YQzEyS1NRX00/edit?usp=sharing"><i>Archaeoglobus, Ferroglobus</i></a>, <a href="https://drive.google.com/file/d/0B7Hl6KFtpv4LRGpsbktvd3ZBeTg/edit?usp=sharing">Rhodothermus</a>, <a href="https://drive.google.com/file/d/0B7Hl6KFtpv4LVmY2Z0NsVmV3ZEk/edit?usp=sharing">Clostridium</a> or <a href="https://drive.google.com/file/d/0B7Hl6KFtpv4LbHl4bXFkTk1DOE0/edit?usp=sharing">Unkown</a>.  
</p><p>
In short, Genbank files with genome annotations are fetched from <a href="http://www.ncbi.nlm.nih.gov/">NCBI</a> (or from somewhere else). 
The genes are blasted all against all (requires computing power) and then clustered together according to the 50-50 rule (can be changed).
The 50-50 rule groups together genes which have 50% similarity in at least 50% of the length.
</p><p>
The resulting clusters are mapped to several databases, such as: <a href="http://geneontology.org">Gene ontology</a>, 
COG and pFam for functional annotation and <a href="http://kegg.jp">KEGG</a> for metabolic pathway analysis.
</p>
</section>
<section>
<h3>Software</h3>
<p>
We use a Java applet based software to generate all kinds of comparative genomics images. 
The software is available at <a href="http://thermusgenes.appspot.com/">http://thermusgenes.appspot.com/</a> 
Make sure <a href="http://java.oracle.com/">Java</a> is installed on your computer.
</p>
<p>
After running the software from the site, 
press the "Import" button (top-left) and select your downloaded genome zip-file.
</p>
<figure style="text-align:center">
<img src="mainscreen.png" width="100%" />
<figcaption>Screenshot from the Pan-Core genomics applet</figcaption>
</figure>
<p>
To get a feeling for the scale of the genomes and number of species we draw a Pan-core chart. 
Go to the "Functions" menu and select "Pan-core", a chart should pop-up.
</p>
<figure style="text-align:center">
<img src="pancore.png" width="100%" />
<figcaption>A standard Pan-core plot</figcaption>
</figure>
<p>
You can also draw a "Relation matrix" (also from the functions menu). 
The blue squares show the number of genes in the genomes (with or without duplications)
</p>
<figure style="text-align:center">
<img src="rmatrix.png" width="100%" />
<figcaption>A relation matrix</figcaption>
</figure>
</section>
<section>
<h3>Assignments</h3>
The pan genome is all genes in all species. The core genome is all the genes which occur in every species.
Ok, just read about it on <a href="http://en.wikipedia.org/wiki/Pan-genome">Wikipedia</a>.
<p>   
<b>(1) How many species are there in the set and what are the core genome size and the pan genome size (the numbers in the gray box in the relation-matrix)?</b>
</p>
Type "16S" or "SSU" to filter the gene table to locate the 16S rRNA.
Select the genes from the top table, right click and press "Show DNA sequences" from the popup menu. 
The popup-window is a minor sequence editor. Save the sequences in the editor to a file	by right clicking and selecting "Export".
Copy the Fasta formated text to an online multiple sequence aligner (such as <a href="http://www.ebi.ac.uk/Tools/msa/muscle/">muscle</a>).
Go back to the sequence editor, delete the old sequences (select and press Del) and import a file with the aligned ones.
If you are using Chrome browser you can create a maximum liklihood phylogenetic tree by running "Draw ML tree" in the popup menu. 
Otherwise just use the "Draw tree" method for a simple Neighbour joining tree or use an online phylogenetic calculation services such as <a href="http://www.phylogeny.fr/version2_cgi/one_task.cgi?task_type=phyml">Phylogeny.fr</a>
You can draw the tree using the online phylogenetic tree drawing site <a href="http://webconnectron.appspot.com/Treedraw.html">Treedraw</a>. 
Simply drag drop the resulting tree in Newick text format on the canvas.
<p>
<b>(2) Draw the 16S rRNA phylogenetic tree</b>
</p>
<figure style="text-align:center">
<img src="16Stree.png" width="100%" />
<figcaption>A 16S rRNA tree for the <i>Thermus</i> strains could look something like this</figcaption>
</figure>
<p>
In the functions menu select "Gene atlas".
Select a reference species in the left table and multiple species to draw in the right one. 
Next select the chromosome (contig or plasmid) to use.
Selecting the GC% color button show an genome atlas of GC%. 
Can you see which ones are possibly related from the GC%?
</p>
<p>
With the genome atlas view open we now look for areas with conserved synteny (non-shuffled gene order). 
In the functions menu select "Gene atlas". 
When the image pops up select the "Breakpoint synteny" radio button.
</p>
<p><i>The genome-atlases are missing a legend for now. The lanes are species in the same order as they occur in the species selection table.</i></p>
<figure style="text-align:center">
<img src="tthb8synt.png" width="100%" />
<figcaption>A genome syteni atlas with <i>Thermus thermophilus HB8</i> as a reference. The gray area facing north is highly conserved in terms of synteny.</figcaption>
</figure>
<p>
Look for conserved areas. You can also check the "Rel" button to see how related the genes are in the area (darker areas less divergent).
Press-drag the area you spot, the genes should now be selected in the gene list. You can use "Crop to selection" in the table popup-menu to only show the selected genes.
</p>
<p>
<b>(3) What kind of genes are in the preseved areas?</b>
</p>
<p>
Look for the columns "Present in" and "Group size" in the gene table. 
Of the selected genes, now only select the ones where these columns show the same number, preferrably the number of species. 
You can sort the table according to these column by pressing the column header.
Now you can make a concatenated genes phylogenetic tree from these selected genes by running "Concatenate tree" from the popup menu. 
</p>
<p>
<b>(4) Show the concatenated genes phylogenetic tree and compare with the 16S, do they differ a lot?</b>
</p>
<p>
Now run "Recomb tree" from the "Functions" menu. 
You get a phylogenetic tree based on synteny information (instead of sequence similarities).
</p>
<p>
<b>(5) Does the topology of the synteny phylogenetic tree differ from the others (16S rRNA, concatenated genes)?</b>
</p>
<p>
Now run the "Genome statistics" from the "Functions" menu.
</p>
<p>
<b>(6) What is the average GC% in the species? Do some strains differ?</b>
</p>
<p>
Draw the "COG chart" from the "Functions" menu. 
This charts show what kind of genes are present in the genomes.
</p>
<figure style="text-align:center">
<img src="spiro_cog.png" width="100%" />
<figcaption>Functional groups (COG) from <i>Spirochaeta</i> strains. 
The termite hindgut bacterium <i>Spirochaeta coccoides</i>, 
although with the smallest genome has double the number of 
Cell division and chromosome partitioning (purple) genes than other species of the genus.
Also, transcriptional genes seem more frequent in the <i>Spirocheata buddy</i> strain than in some others.
</figcaption>
</figure>
<p>
Search for "flagella" and "S-layer" in the gene filter/search box. 
By scrolling to the right in the gene table you can see which strains contain specific genes (colored green).
</p>
<p>
<b>(7) Which bacteria strains have (possibly) the ability to move, containing flagellar genes. Which have a possible sessile lifestyle (S-layer)?</b>
</p>
<p>
<b>(8) If you have the <i>Thermus</i> data, look for GO id GO:0009411 (UV response) in the function table and see which strains are deeply submerged and which ones live on the surface.
In the <i>Spirochaeta</i> data you can search for rod-shape determining proteins to see which strains are potentially rod-shaped.
Look for cytoskeletal proteins and the strains containing them.
</b>
</p>
<p>
Select a pathway from the combo box below the gene table. 
Fetch some common pathway (such as Vitamin B6).
Sort the function table with the "Function coverage" column. 
Double-click some GO-ids to show genes with that annotation in the gene table. Gather some KO-ids.
Now go to <a href="http://pathways.embl.de">iPath2</a>. 
Press the "customize" button. You can paste your KO-ids (or EC-isd) in the "element selection" window.
For example "E6.3.4.15 #ff0000", "K00831 #0000ff", the latter "#" string being the color you want to use to highlight the pathway.
</p>
<p>
Look for the "chromosomal replication initiator protein" (dnaA). This protein signals the start of the DNA replication.
Highlight it and do a "Genome atlas". Be careful to press esc on the gene table before to un-crop it. THe gene atlas only shows the genes visible in the table.
Select some interesting strains, like all the <i>Spirochaeta</i> or the <i>Thermus thermophiluses</i>.
Press the "Synteny gradient" button. Now you see a rainbow colored circles indicating the genome order compared to the reference strain.
</p>
<figure style="text-align:center">
<img src="syntrainbow.png" width="100%" />
<figcaption>
Synteny gradient image from <i>Thermus</i>.
</figcaption>
</figure>
</section>
</div>
</body>
</html>