<!DOCTYPE html><html><head>
<title>Pan-core and functional genome navigator</title>
    <link rel="stylesheet" type="text/css" href="pancore.css">
    <link href="http://fonts.googleapis.com/css?family=Headland+One|Life+Savers:400,700" rel="stylesheet" type="text/css">
</head>
<body>
<table cellspacing="20px" style="font-family: 'Life Savers', cursive;"> <!--'Headland One', serif;-->
<tbody><tr><td colspan="3">
      <h2>Pan-core and functional genome navigator</h2>
	<h3><i>Sigmar Karl Stef√°nsson</i></h3>
      </td></tr>
    <tr><td>
    <h3>Introduction</h3>
Comparative genomics. We present a software suite for navigating and comparing the genomes of multiple related microbial species. The package is written in Java and is accessible on the web as a Java applet on the URL <a href="http://thermusgenes.appspot.com/">http://thermusgenes.appspot.com/</a>. <p><figure style="text-align:center">
<img src="mainscreen.png" width="300px" height="200px" />
<figcaption>The PanCoreNav mainscreen</figcaption>
</figure>
The mainscreen is split into two main tables, the gene table on top and the function table below. The tables are linked so be double clicking on an entry the relevant row in the other table is highlighted. <p>
A search textbox accompanies both tables for easy search and filtering.<p>
<b>Relation matrix</b><p>
The 'Relation matrix' in the Functions menu gives you an image showing number of related genes between all species.
<figure style="text-align:center">
<img src="rel.png" width="300px" height="200px" />
<figcaption>Relation matrix</figcaption>
</figure>
<b>Gene frequency distribution</b><p>
The gene frequency distribution shows the frequency of occurance of the genes. The core genome is indicated by the last value in the dot plot.
<figure style="text-align:center">
<img src="freqd.png" width="300px" height="200px" />
<figcaption>Gene frequency distribution (the graph shows that the core genome is little less than 1000 genes and single occuring genes are close to 4000)</figcaption>
</figure>
    </td><td>
    <h3>Prepare your data</h3>
You will need fasta files containing the contigs of each of your species (or whichever level you would loke to explore the pan-genome, we will use species hereafter) and named after the species. These file are drag-dropped into a special online fasta manipulator software <a href="http://funblastic.appspot.com/">http://funblastic.appspot.com/</a>. Here you join the fasta files for each species into one, inserting the name of the species in front of the contig name. Highlight all the fasta files you want to join and select "Join" from the right click popup menu. Now highligh only the result file, which should have the ending "_joined.*". <p> The Prodigal <cite>http://prodigal.google.com</cite> software is used for gene finding. From the same popupmenu run the function "Prodigal". When the progress bar stops you should have two new fasta files containing the name of the original file along with "prodigal" and ".fsa" for protein fasta file and ".fna" for nucletide fasta results file. <p> Now we need to find out which genes are related to each other. You select the amino acid genes fasta file, the one with the ending ".fsa" and run "Create database" from the popup menu. This creates a blast database from you fasta file and inserts it into the blast database table on the right.<p>
Genome annotations can be very time consuming / CPU intensive. If you have already annotated your genomes you can skip these steps and drag-drop genbank annotation files into the table (from <a href="http://rast.nmpdr.org/">RAST</a> annotatino pipeline for instance). They will be split into corresponding fasta files. Also it is possible to fetch genbank files directly form NCBI ftp site according to some search term using "NCBI fetch" function.<p>
Using the 50-50 rule 
  </td><td>
    <h3>Basic functionality</h3>
The "Pan-core" action in the Functions menu gives you a standard Pan-core plot for you species. <p>
<figure style="text-align:center">
<img src="pancore.png" width="300px" height="200px" />
<figcaption>Standard Pan-core plot</figcaption>
</figure><p>
You can change the order of species by drag droping in the species selection table<p>
<b>XY plot</b><p>
The 'XY plot' function in the Functions menu draw a dotplot between two selected genomes
<figure style="text-align:center">
<table cellspacing="5"><tr><td>
<img src="xy1.png" width="150px" height="150px" /></td><td>
<img src="xy2.png" width="150px" height="150px" /></td>
</tr></table>
<figcaption>Genome XY plot (before and after filtering the Gene table)</figcaption>
</figure><p>
By filtering the gene table in the main screen it is possible to select which genes are visible in the XY plot. For example by sorting the gene table by '# locus' you can filter out genes with multiple copies (such as transposons). <p> The XY plot offers different color schemes. For example colors can be selected by the gene GC content or the gene total abundance to help spotting genetic islands. <p>
Selection in the XY view propogates to the gene table (use double click to unselect).
<p><b>Synteny view</b><p>
The synteny view shows the position of the genes in relation to other species.
<figure style="text-align:center">
<img src="synteny.png" width="300px" height="200px" />
<figcaption>Standard Pan-core plot</figcaption>
</figure><p>
    </td><td rowspan="3">
<ins class="adsbygoogle"
     style="display:inline-block;width:160px;height:600px"
     data-ad-client="ca-pub-7204381538404733"
     data-ad-slot="8243040944"></ins>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
</td></tr>
  <tr><td colspan="3">
  <h3>Gene viewer</h3>
Pressing the "Neighborhood" button in the bottom row gives you a view of the gene neighbourhood for each species. You can manually align the genes (on genome level, not on sequence level) by using ... function. Different color schemes show possible horizontal gene transfer events or genomic islands.<p>
<figure style="text-align:center">
<img src="genes.png" width="800px" />
<figcaption>Gene neighbourhood view (with the abundance color scheme activated, green genes have low abundance)</figcaption>
</figure><p>
g
    </td></tr>
    <tr><td colspan="2">
    <h3>Sequence viewer / editor</h3>
To visualize and edit the sequences. Select "Show sequences" from the popup-menu after selecting your genes of interest.
<figure style="text-align:center">
<img src="seqview.png" width="600px" />
<figcaption>A screenshot from the online sequenceviewer (<a href=http://webfasta.appspot.com/">http://webfasta.appspot.com/</a>)</figcaption>
</figure><p>
<br>
<ins class="adsbygoogle"
     style="display:inline-block;width:728px;height:90px"
     data-ad-client="ca-pub-7204381538404733"
     data-ad-slot="6766307742"></ins>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
      </td><td>
	<h3>Draw phylogenetic trees</h3>
You can create phylogenetic tree based on the genes in your dataset. For instance you can select the 16S rRNA, right click and select "Draw tree" using neighbour joining or "Draw ML tree" using FastTree method. Also you can draw tree based on multiple concatenated genes or Majority-rule consensus between them. The treedrawing software used is the web-based <a href="http://webconnectron.appspot.com/Treedraw.html">Treedraw (http://webconnectron.appspot.com/Treedraw.html)</a>. <p>
<figure style="text-align:center">
<img src="tree.png" width="300px" />
<figcaption>Phylogenetic tree based on concatenated genes from the core genome</figcaption>
</figure><p>
<b>Genome wide phylogeny</b><p>
<i>Synteny</i><p>
You can also draw phylogenetic trees based on synteny. A distance matrix is created based on the number of adjacent related genes between two species.
<figure style="text-align:center">
<img src="syntree.png" width="300px" />
<figcaption>Phylogenetic tree based on genome synteny</figcaption>
</figure><p>
<i>Presence-absense</i><p>
Presense absense trees are phylogenetic trees based on binary data, e.g. the presense or absense of genes. It is also a distance matrix based method.
<figure style="text-align:center">
<img src="presabstree.png" width="300px" />
<figcaption>Phylogenetic tree based on gene presense absense (the topology of the genome wide trees agree quite well with the trees based on sequence alignment)</figcaption>
</figure><p>
    </td></tr>
     <tr><td colspan="4">
    <h3>Discussion</h3>
The need for viewing and navigating the genomes of bacterial species compared related species.
  </td></tr>
    <tr><td colspan="4" align="center">
      <font size="2"><a href="mailto:sigmarkarl@gmail.com">sigmarkarl@gmail.com</a> Disclaimer: This is a joke</font>
    </td></tr>
	</tbody></table>

</body></html>
