<!DOCTYPE html><html><head>
<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<title>Pan-core gene and function navigator</title>
    <link rel="stylesheet" type="text/css" href="pancore.css">
    <link href="http://fonts.googleapis.com/css?family=Headland+One|Life+Savers:400,700" rel="stylesheet" type="text/css">
    <script async src="http://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
</head>
<body style="text-align : justify; font-family: 'Life Savers', cursive; font-size: 12px">
<div style="text-align: center" class="do-not-print">
<ins class="adsbygoogle"
     style="display:inline-block;width:728px;height:90px"
     data-ad-client="ca-pub-7204381538404733"
     data-ad-slot="8470040143"></ins>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
</div>
<h2>Pan-core gene and function navigator</h2><div class="g-plusone" data-annotation="inline" data-width="300"></div>
<h3><i>Sigmar Karl Stef√°nsson</i></h3>
	<script type="text/javascript">
  (function() {
    var po = document.createElement('script'); po.type = 'text/javascript'; po.async = true;
    po.src = 'https://apis.google.com/js/plusone.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(po, s);
  })();
</script>
<div class="article">

<h3>Introduction</h3>
We present a comparative genomics software suite for navigating and comparing the genomes of multiple related microbial species. 
The package is written in Java and is accessible on the web as a Java applet on the URL <a href="http://thermusgenes.appspot.com/">http://thermusgenes.appspot.com/</a>. <p>
You can try out the software with already prepared data, such as from <a href="allspiro.zip"><i>Spirochaeta</i></a> sequences.
<figure style="text-align:center">
<img src="mainscreen.png" width="90%" />
<figcaption>The PanNavigator mainscreen</figcaption>
</figure>
The mainscreen is split into two main tables, the <b>gene table</b> on top and the <b>function table</b> below. The tables are linked so be double clicking on an entry the relevant row in the other table is highlighted.
A search textbox accompanies both tables for easy search and filtering.<p>
<h3>Prepare your data</h3>
You will need fasta files containing the contigs of each of your species (or whichever level you would like to explore, we will refer to species hereafter) 
and named after the species. These file are drag-dropped into a special online fasta manipulator software <a href="http://funblastic.appspot.com/">http://funblastic.appspot.com/</a>. 
Here you join the fasta files for each species into one, inserting the name of the species in front of the contig name. 
Highlight all the fasta files you want to join and select "Join" from the right click popup menu. Now highligh only the result file, which should have the ending "_joined.*". 
<p><cite href="http://prodigal.google.com">The Prodigal</cite> software is used for gene finding. From the same popupmenu run the function "Prodigal". When the progress bar stops you should have two new fasta files containing the name of the original file along with "prodigal" and ".fsa" for protein fasta file and ".fna" for nucletide fasta results file. <p> Now we need to find out which genes are related to each other. You select the amino acid genes fasta file, the one with the ending ".fsa" and run "Create database" from the popup menu. This creates a blast database from you fasta file and inserts it into the blast database table on the right.<p>
Genome annotations can be very time consuming / CPU intensive. If you have already annotated your genomes you can skip these steps and drag-drop genbank annotation files into the table (from <a href="http://rast.nmpdr.org/">RAST</a> annotatino pipeline for instance). They will be split into corresponding fasta files. Also it is possible to fetch genbank files directly form NCBI ftp site according to some search term using "NCBI fetch" function.<p>
The 50-50 rule is used to group the genes together. That is, at least 50% of the sequence length must have at least 50% sequence identity.
<figure style="text-align:center">
<img src="funblastic.png" width="90%" />
<figcaption>A screenshot from the funblastic site</figcaption>
</figure><p>
Now you have a new list of protein fasta files containing the all the sequences of a gene group. These files represent the pan genome. The subset of these files with the maximum number of sequences is the core genome. 
To be able to concatenate and make phylogenetic trees these need to be aligned. Funblastic gives you two options, alignment with muscle or clustalw. Now you should have all the files needed for import to the PanNavigator software, 
one nucleotide fasta file containing the contigs from all the microbial organism under investigation and a protein fasta file for each of the genes.
<p>All the results, for the alignment, clustering and annotation are kept in fasta format for readability and portability.

 <h3>Basic functionality</h3>
    <b>Relation matrix</b><p>
The 'Relation matrix' in the Functions menu gives you an image showing number of related genes between all species.
<figure style="text-align:center">
<img src="rel.png" width="90%" />
<figcaption>Relation matrix</figcaption>
</figure>
<b>Gene frequency distribution</b><p>
The gene frequency distribution shows the frequency of occurance of the genes. The core genome is indicated by the last value in the dot plot.
<figure style="text-align:center">
<img src="freqd.png" width="90%" />
<figcaption>Gene frequency distribution (the graph shows that the core genome is little less than 1000 genes and single occuring genes are close to 4000)</figcaption>
</figure><p>
<b>GC% histogram</b><p>
The GC histogram shows histogram of average GC% for all the gene families.
<figure style="text-align:center">
<img src="gcp.png" width="90%" />
<figcaption>Gene GC% histogram</figcaption>
</figure><p>
The GC content averages around 67% for the <i>Thermus</i> species in the example data<p>
The "Pan-core" action in the Functions menu gives you a standard Pan-core plot for you species. <p>
<figure style="text-align:center">
<img src="pancore.png" width="90%" />
<figcaption>Standard Pan-core plot</figcaption>
</figure><p>
You can change the order of species by drag droping in the species selection table<p>
<b>XY plot</b><p>
The 'XY plot' function in the Functions menu draw a dotplot between two selected genomes
<figure style="text-align:center">
<img src="xy1.png" width="45%" />
<img src="xy2.png" width="45%" />
<figcaption>Genome XY plot (before and after filtering the Gene table)</figcaption>
</figure><p>
By filtering the gene table in the main screen it is possible to select which genes are visible in the XY plot. For example by sorting the gene table by '# locus' you can filter out genes with multiple copies (such as transposons). <p> The XY plot offers different color schemes. For example colors can be selected by the gene GC content or the gene total abundance to help spotting genetic islands. <p>
Selection in the XY view propogates to the gene table (use double click to unselect).
<p><b>Synteny view</b><p>
The PanNavigator software genome alignment is based on the genes. The synteny view shows the position of the genes in relation to other species.
<figure style="text-align:center">
<img src="synteni.png" width="90%" />
<figcaption>Synteny plot showing the selection of the possibly highly expressed 'metal dependent phosphohydrolase' in <i>Thermus</i> species</figcaption>
</figure><p>
Selecting one gene in the gene table and visualizing in the synteny viewer gives an indication of gene expression level if there are multiple copies of the genes (paralogs).
The synteny view is also useful in spotting which genes lie adjacent to each other as it is linked to selection in the gene table
<figure style="text-align:center">
<img src="synteny.png" width="90%" />
<figcaption>Single (or few) gene synteny plot</figcaption>
</figure><p>
<b>Sorting in the Gene table</b><p>
The gene table is number of useful columns to sort by. In the synteny example above we saw a gene with multiple copies. Sorting by '# locus' enables you to filter out those genes.
<p><i>Sharing number</i><p>
Each permutation of gene sharing is given a number, so by sorting by 'Sharing number' groups together genes who are shared by the same species. 

<!-- div class="do-not-print">
<ins class="adsbygoogle"
     style="display:inline-block;width:160px;height:600px"
     data-ad-client="ca-pub-7204381538404733"
     data-ad-slot="3900239744"></ins>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
</div-->

</div>

<div>
<h3>Genome atlas</h3>
The gene atlas uses one genome as a reference and shows the presense or absense of the genes (in that order) in other species. The genes are colored by default by resemblance. Also it is possible to color by GC content.
This view can be used to spot highly preserved areas or areas with high variability.
<figure style="text-align:center">
<img src="scotoductusSA01.png" width="45%" />
<img src="scotoductusSA01_gc.png" width="45%" />
<img src="oshimaiJL2.png" width="45%" />
<img src="thermophilusJL18.png" width="45%" />
<figcaption>The upper row shows genome atlases for the <i>Thermus scotoductus SA_01</i> 
with two color palettes. The first shows resemblance color scheme. The dark circles are <i>Thermus scotoductus</i> 
and the outer dark circle the related <i>Thermus antranikianii</i>. The innermost low resemblance area are species form different genera, 
the <i>Meiothermus, Marinithermus and Oceanitherus</i>. The atlas on the right has the same reference genome with GC% color scheme. 
The <i>Thermus scotoductus</i> strains have more resemblance with the strains from other genera than speceis from same genera in terms of GC%. 
The dark area in all the genome atlases represents a highly preserved locus containing mostly 50S and 30S related genes. 
The single unrelated <i>Thermus</i> species (the lightgray strip) in all the resemblance atlases is the <i>Thermus filiformis</i> from New Zealand</figcaption>
</figure><p> 
  <h3>Gene viewer</h3>
Pressing the "Neighborhood" button in the bottom row gives you a view of the gene neighbourhood for each species. You can manually align the genes (on genome level, not on sequence level) by using ... function. Different color schemes show possible horizontal gene transfer events or genomic islands.<p>
<figure style="text-align:center">
<img src="genes.png" width="90%" />
<figcaption>Gene neighbourhood view (with the abundance color scheme activated, green genes have low abundance)</figcaption>
</figure><p>
The blocks view is for gene level alignment. Checking the "Show in-betweens" show the length of the areas between the genes (annotations) non-annotated areas indicating possible sequences of interest like repeast or CRISP sequences.<p>

<b>Gene sorter</b><p>
The gene sorter view can be useful when manually connecting contigs. Also usig the different color schemes it is easy to spot genomic islands.
<figure style="text-align:center">
<img src="genesorter.png" width="90%" />
<figcaption>Gene sorter view. The 'Sharing number' color scheme shown here shows alternating colors when the permutation of sharing among species changes. The three red arrows show possible usecases of the gene sorter view. By selecting the large highly preserved area the <b>genetable</b> shows that it contains mostly RNA genes. Other arrow shows how shifting to GC% color scheme, the region which is apparent in some <i>Thermus scotoductus</i> but missing in other has different GC% than the surroundings. This indicates that the region could be a chunk of horizontally transferred genes. As to be expected, looking at the gene table, they are mostly hypothetical genes. The top arrow gives an indication of how contigs can be connected in related species (although this method can never be 100% accurate)</figcaption>
</figure><p>

<b>Connecting contigs</b><p>
The gene viewer can also be used to connect contigs based on gene content in the contigs. 
This can be done manually in the "Neighbourhood" view or automatically with the "Connect contigs" function.
The Newbler assembler from the FLX 454 software package gives a contig connection graph which can be used to help.
To visualize the graph select "Load contig graph" from the Functions menu. 
</div>

<div class="article">
    <h3>Sequence viewer / editor</h3>
To visualize and edit the sequences. Select "Show sequences" from the popup-menu after selecting your genes of interest.
<figure style="text-align:center">
<img src="seqview.png" width="90%" />
<figcaption>A screenshot from the online sequenceviewer (<a href=http://webfasta.appspot.com/">http://webfasta.appspot.com/</a>)</figcaption>
</figure><p>
<b>Gene gain/loss</b><p>
It is possible to view gene/function gain and loss according to some selected phylogeny.
Right click and select "Gene gain-loss" functino from the popup menu.
<figure style="text-align:center">
<img src="genes.png" width="90%" />
<figcaption>Circular gene gain/loss tree</figcaption>
</figure><p>
<b>Species / Function matrix</b><p>
A matrix showing the occurance of Metabolic pathways in each species
	<h3>Draw phylogenetic trees</h3>
You can create phylogenetic tree based on the genes in your dataset. For instance you can select the 16S rRNA, right click and select "Draw tree" using neighbour joining or "Draw ML tree" using FastTree method. Also you can draw tree based on multiple concatenated genes or Majority-rule consensus between them. The treedrawing software used is the web-based <a href="http://webconnectron.appspot.com/Treedraw.html">Treedraw (http://webconnectron.appspot.com/Treedraw.html)</a>. <p>
<figure style="text-align:center">
<img src="tree.png" width="90%" />
<figcaption>Phylogenetic tree based on concatenated genes from the core genome</figcaption>
</figure><p>
<b>Genome wide phylogeny</b><p>
<i>Synteny</i><p>
You can also draw phylogenetic trees based on synteny. A distance matrix is created based on the number of adjacent related genes between two species.
<figure style="text-align:center">
<img src="syntree.png" width="90%" />
<figcaption>Phylogenetic tree based on genome synteny</figcaption>
</figure><p>
<i>Presence-absense</i><p>
Presense absense trees are phylogenetic trees based on binary data, e.g. the presense or absense of genes. It is also a distance matrix based method.
<figure style="text-align:center">
<img src="presabstree.png" width="90%" />
<figcaption>Phylogenetic tree based on gene presense absense (the topology of the genome wide trees agree quite well with the trees based on sequence alignment)</figcaption>
</figure><p>
</div>

<div>
<h3>Navigating metabolic pathways</h3><p>
Using chemical measurements we can verify and compare the presence and activity of metabolic pathways. For a simple case using corresponance analysis of 16S taxonomic data compared to annotations of 'water', 'soil' and 'biomass'.
<figure style="text-align:center">
<img src="vist1.png" width="90%" />
<figcaption>Corresponance analysis of taxonomic vs niche annotation</figcaption>
</figure><p>
    <h3>Discussion</h3>
The need for viewing and navigating the genomes of bacterial species compared related species.
<font size="2"><a href="mailto:sigmarkarl@gmail.com">sigmarkarl@gmail.com</a> | <a href="https://plus.google.com/115473796046446928448">Find us on Google+</a></font>
</div>
</body>
</html>