<!DOCTYPE html>
<html><head>
<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<title>Pan genome of the Thermus species</title>
    <link rel="stylesheet" type="text/css" href="pancore.css">
    <link href="http://fonts.googleapis.com/css?family=Headland+One|Life+Savers:400,700" rel="stylesheet" type="text/css">
    <!--script async src="http://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script-->
</head>
<body style="text-align : justify; font-family: 'Life Savers', cursive; font-size: 12px">
<!--div style="text-align: center" class="do-not-print">
<ins class="adsbygoogle"
     style="display:inline-block;width:728px;height:90px"
     data-ad-client="ca-pub-7204381538404733"
     data-ad-slot="8470040143"></ins>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
</div-->

<div class="article">
<div style="-webkit-column-span:all;-moz-column-span:all;column-span:all">
<h2>Pan genome of the <i>Thermus</i> species</h2><br>
<div class="g-plusone" data-annotation="inline" data-width="300"></div>
<h3><i>Stefánsson, Sigmar K., Björnsdóttir, Snædís, Friðjónsson, Ólafur H., Pétursdóttir, S.P., Hreggviðsson, Guðmundur Ó.</i></h3>
<!--script type="text/javascript">
  (function() {
    var po = document.createElement('script'); po.type = 'text/javascript'; po.async = true;
    po.src = 'https://apis.google.com/js/plusone.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(po, s);
  })();
</script-->
</div>
<section>
<b>Abstract</b>
<p>
The phylogenetic relationship between known Thermus species in geographic context was studied. 
The Thermus genera are a group of thermophilic bacteria residing in hot springs. 
These special living conditions can be thought of as islands in the ecological sense. 
Sample locations were compared to environmental variables such as pH and temperature to determine which more influences species diversity, 
geographic separation or chemical composition in hot springs.
</p>
</section>
<section>
<b>Introduction</b><p>
<p>Bacteria of the genus Thermus were first isolated from hot springs in Yellowstone National Park, USA.  These bacteria were described as Thermus aquaticus by Brock and Freeze in 1969.  The species is well recognized as the type strain, YT1, is the source of the Taq polymerase, an enzyme that has been of a great biotechnological importance (Brock, 1997).  Soon after the discovery of T. aquaticus, isolations of similar strains were reported from geothermal areas (Oshima and Imahori 1971; Pask-Hughes and Williams 1977; Kristjansson and Alfredsson 1983; Munster et al 1986) as well as from man-made hot water systems and thermally polluted streams (Ramaley and Hixson 1970; Brock and Boylen 1973; Pask-Hughes and Williams 1975; Degryse et al 1978).  The taxonomic position of Thermus was unknown at this time but a phylogenetic relationship with the genus Deinococcus was later realized (Hensel et al 1986).  The two genera are now classified in a common phylum, “Deinococcus-Thermus” and constitute the type genera of family Thermaceae of order Thermales (da Costa and Rainey 2001) and of family Deinococcaceae of order Deinococcales (Battista and Rainey 2001), respectively.  The order Deinococcales is renowned for the high radiation resistance of some strains and contains both mesophiles and thermophiles whereas all members of the order Thermales are thermophiles.</p>
<p class="tab">Thermales strains are rod-shaped, non-spore-forming, predominantly heterotrophic and grow by aerobic respiration although other electron acceptors such as nitrate, nitrite, sulfur and metals have been reported (Kieft et al 1999; da Costa and Rainey 2001).  Different temperature optima for growth have been observed within the group.  Thermus strains grow optimally at about 65-75°C (Nobre et al 1996).  A second genus, Meiothermus, was proposed to comprise strains that grow at lower temperatures, optimally at 45-65°C (Nobre et al 1996; Albuquerque et al 2010).  Meiothermus includes the species M. ruber, M. silvanus and M. chliarophilus that were originally described within the genus Thermus.  In recent years, four new genera have been created in family Thermaceae to accommodate representatives that were isolated from differently located deep-sea hydrothermal habitats.  The genera, Marinithermus (Sako et al 2003), Oceanithermus (Miroshnichenko et al 2003a), Rhabdothermus (Steinsbu et al 2011) and Vulcanithermus (Miroshnichenko et al 2003b) encompass thermophilic strains with temperature optima of about 60-70°C.  The new isolates are phenotypically similar to members of Thermus and Meiothermus although some are microaerophilic and growth on molecular hydrogen has been reported for Oceanithermus and Vulcanithermus (Miroshnichenko et al 2003a, 2003b).  Distinctively, the deep-sea strains require salt for growth and grow optimally at about 3% NaCl.</p>
<p class="tab">Strains of the genus Thermus are highly variable in terms of certain phenotypic characters but do form tight phylogenetic clusters (da Costa and Rainey 2001; Kristjansson and Hreggvidsson 2004; Hreggvidsson et al 2006).  In addition to T. aquaticus the genus currently entails ten validly described species.  These are T. antranikianii (Chung et al 2000), T. arciformis (Zhang et al 2010), T. brockianus (Williams et al 1995), T. composti (Vajna et al 2012), T. filiformis (Hudson et al 1987), T. igniterrae (Chung et al 2000). T. islandicus (Bjornsdottir et al 2009), T. oshimai (Williams et al 1996), T. scotoductus (Kristjansson et al 1994) and T. thermophilus (Manaia et al 1994).  In addition, three proposed species, „T. kawarayensis“ (Kurosawa et al 2005), „T. rehai“ (Lin et al 2002) and „T. yunnanensis“ (Gong et al 2005) have not been validated.  Strains from all these species have been isolated from hydrothermal vents except for the newest member that derives from hot compost (Vajna et al 2012).  Disparate geographical distribution has been observed for Thermus species.  Some appear to be restricted to certain areas, such as T. aquaticus and T. filiformis, which have only been isolated from hot springs in North America and New Zealand, respectively (Brock and Freeze 1969; Hudson et al 1987; da Costa and Rainey 2001).  Others, such as T. brockianus, T. oshimai, T. scotoductus and T. thermophilus have been isolated from differently located sites (Kristjansson et al 1994; Manaia et al 1994; Williams et al 1995, 1996; Hreggvidsson et al 2006).  Geographical isolation and adaptation to certain habitats might have been important for species divergence within Thermus.  Growth in saline environments is, however, the only character that can be clearly related to geographical distribution.  T. thermophilus is the only Thermus species that has been isolated from coastal and marine locations (Kristjansson et al 1986; Manaia and da Costa 1991; Marteinsson et al 1999).  A comparison of Icelandic strains further indicated that pH and conductivity affect Thermus distribution (Hreggvidsson et al 2006).  Habitat preferences within the genus are nevertheless largely unexplained and probably result from subtle physiological differences (Hreggvidsson et al 2011).  We have undertaken extensive sequencing efforts and are currently examining whether indications of adaptive traits in Thermus can be found by comparative genomics.</p>
<p class="tab">The increasing availability of molecular data allows the examination of microbial biogeography based on environmental distribution of gene sequences.  For this purpose and to expand the knowledge on Thermus distribution, we extracted and reviewed all available 16S rRNA gene sequences that belong to the genus from the SILVA SSUParc database.  Here, we present the global distribution of Thermus based on this data, which derives predominantly from culture-independent studies.  We also present a framework for managing and analyzing such data that could benefit studies of other taxonomic groups.  Furthermore, we relate the emerging picture of Thermus distribution with the taxonomy of the genus and discuss potential new groups.  We also examine phylogenetic relationships by including preliminary whole genome sequence data of several Thermus strains.</p>
</section>
<section>
<b>Preparation of the data</b>
<p>The strains available online were already annotated. For the MATIS species we used Prodigal 2.6 for gene identification and GenePRIMP for further refinement. All the genes were clustered together using the 50-50 rule and the sequences in each group aligned together using the Muscle multiple sequence aligner.
The remaining clusters without annotation were blasted against the nr database.</p>
</section>
<section>
<b><i>Thermus</i> genome statistics</b>
<p>We compare 39 whole genome sequenced strains from the <i>Thermacae</i> family with focus on the <i>Thermus</i> generera. Of those, 32 are <i>Thermus</i> species. 18 of them are sequenced at <a href="http://www.matis.is/">MATIS</a> and are draft genomes. 11 are available from <a href="http://ncbi.nlm.nih.gov">NCBI</a> of which 4 are draft genomes. Also we used 3 draft sequences from the <a href="http://genome.jgi.doe.gov/">Join Genome Institue</a> at DOE.
The other generas, <i>Meiothermus</i>, <i>Marinithermus</i> and <i>Ocenithermus</i> are kept for phylogenetic tree rooting and comparision. 3 of Meiothermus strains are draft genomes, the other completed, in total 7 strains.
</p>
<p>In terms of gene number, there is no notable difference between the <i>Thermus</i> strains, on average around 2400 genes. The genome GC% is lower in the <i>Thermus scotoductus</i> and the closely related <i>Thermus antranikianii</i> (65.0% compared to the average of 68.5% for the other species).
The <i>Thermus brockianus</i> and <i>Thermus eggertsoni</i> also have similar genome GC% different from the rest (66.9%).
</p>
<figure style="text-align:center">
<img src="gsize.png" width="100%" />
<!--div id="gsizechart_div" style="width: 100%; height: 40em"></div-->
<figcaption><i>Thermus</i> genome sizes, the average being 2400 genes</figcaption>
</figure>

<figure style="text-align:center">
<img src="gc.png" width="100%" />
<!--div id="gcchart_div" style="width: 100%; height: 30em"></div-->
<figcaption><i>Thermus</i> GC%</figcaption>
</figure>

<table style="font-size: 0.5em"><tr><td>
Marinithermus_hydrothermalis_DSM_14884_uid65783</td></tr><tr><td>
Meiothermus_ruber_DSM_1279_uid198526</td></tr><tr><td>
Meiothermus_ruber_DSM_1279_uid46661</td></tr><tr><td>
Meiothermus_ruber_H328_uid199487</td></tr><tr><td>
Meiothermus_silvanus_DSM_9946_uid49485</td></tr><tr><td>
Meiothermus_timidus_DSM_17022_uid199231</td></tr><tr><td>
Oceanithermus_profundus_DSM_14977_uid60855</td></tr><tr><td>
Thermus_CCB_US3_UF1_uid81197</td></tr><tr><td>
Thermus_RLM_uid156767</td></tr><tr><td>
Thermus_antranikianii_DSM_12462</td></tr><tr><td>
Thermus_antranikianii_MATIS</td></tr><tr><td>
Thermus_aquaticus_Y51MC23_uid55053</td></tr><tr><td>
Thermus_arciformis_MATIS</td></tr><tr><td>
Thermus_brockianus_104_MATIS</td></tr><tr><td>
Thermus_brockianus_MATIS</td></tr><tr><td>
Thermus_eggertsoni_MATIS</td></tr><tr><td>
Thermus_filiformis_MATIS</td></tr><tr><td>
Thermus_igniterrae_ATCC_700962</td></tr><tr><td>
Thermus_igniterrae_MATIS</td></tr><tr><td>
Thermus_islandicus_DSM_21543</td></tr><tr><td>
Thermus_islandicus_MATIS</td></tr><tr><td>
Thermus_kawarayensis_MATIS</td></tr><tr><td>
Thermus_oshimai_DSM_12092</td></tr><tr><td>
Thermus_oshimai_JL_2_uid178948</td></tr><tr><td>
Thermus_oshimai_MATIS</td></tr><tr><td>
Thermus_scotoductus_1572_MATIS</td></tr><tr><td>
Thermus_scotoductus_2101_MATIS</td></tr><tr><td>
Thermus_scotoductus_2127_MATIS</td></tr><tr><td>
Thermus_scotoductus_252_MATIS</td></tr><tr><td>
Thermus_scotoductus_346_MATIS</td></tr><tr><td>
Thermus_scotoductus_4063_MATIS</td></tr><tr><td>
Thermus_scotoductus_DSM_8553</td></tr><tr><td>
Thermus_scotoductus_SA_01_uid62273</td></tr><tr><td>
Thermus_thermophilus_791_MATIS</td></tr><tr><td>
Thermus_thermophilus_ATCC_33923</td></tr><tr><td>
Thermus_thermophilus_HB27_uid58033</td></tr><tr><td>
Thermus_thermophilus_HB8_uid58223</td></tr><tr><td>
Thermus_thermophilus_JL_18_uid162129</td></tr><tr><td>
Thermus_thermophilus_SG0_5JP17_16_uid159537</td></tr>
</table>

<p>The <i>Thermus</i> species are similar in terms of functional annotions. The small groups, Defence mechanism and Secondary metabolism COG groups, being the most variable.</p>

<div style="-webkit-column-span:all;-moz-column-span:all;column-span:all">
<figure style="text-align:center">
<img src="cog.png" width="100%" />
<!--div id="cogchart_div" style="width: 100%; height: 50em"></div-->
<figcaption><i>Thermus</i> COG chart</figcaption>
</figure>
</div>

</section><section>
<b><i>Thermus</i> pan-core genome</b>
<p>Using all the genes from the 32 whole genomes sequenced <i>Thermus</i> strains, the pan genome is close to 4600 genes.
The core genome is around 1100 genes.</p>
<figure style="text-align:center">
<img src="pc.png" width="100%" />
<!--div id="chart_div" style="width: 100%; height: 30em"></div-->
<!--img src="pancorethermus.png" width="100%" /> -->
<figcaption>A phylogenetic tree based on 16S rRNA</figcaption>
</figure>

</section>
<section>
<b>Phylogeny of the <i>Thermus</i> species</b>
<p>The phylogeny of the Thermus species varies depending in which genes are used in constructing the phylogeny.
We compare the conventioal 16S rRNA phylogeny to multiple gene phylogenies.</p>  
<figure style="text-align:center">
<img src="16Stree.png" width="100%" />
<figcaption>A phylogenetic tree based on 16S rRNA</figcaption>
</figure>

<p>To select the phylogenetically most relevant genes we draw genome atlas with <i>Thermus thermophilus HB8</i> as the reference species.</p>
<figure style="text-align:center">
<img src="tthb8atlas.png" width="100%" />
<figcaption>A Genome relation atlas with <i>Thermus thermophilus HB8</i> as a reference</figcaption>
</figure>

<figure style="text-align:center">
<img src="tthb8synt.png" width="100%" />
<figcaption>A genome syteni atlas with <i>Thermus thermophilus HB8</i> as a reference</figcaption>
</figure>

<p>Eyeballing the genome relation atlas shows the New Zealandic <i>Thermus filiformis</i> is distantly related to other <i>Thermuses</i> (the light greyish circle in the middle of the atlas). 
Contraty to the 16S tree the <i>Thermus arciformis</i> and <i>Thermus kawarayensis</i> to a lesser extent are genomewise more related to the <i>Thermus thermophilus</i> than to <i>Thermus islandicus</i> (the darkish lines on either side of the <i>Thermus filiformis</i>)</p>

<p>There are two areas both highly preserved in terms of both synteny and mutation. Both areas are also reported with preserved synteny in [The genome sequence of the extreme thermophile <i>Thermus thermophilus</i>] as a cluster encoding ribosomal proteins (TTC1316-1329) and a NADH-quinone oxidoreductase gene cluster (TTC1907-1920). 
The darkish arc pointing north in the Genome relation atlas is a cluster of 34 ribosomal genes (50S and 30S) and higly preserved throughout the <i>Thermacaea</i> family. 
Evidently from the accompanying synteny atlas the cluster is also preserved in terms of synteny. Concatenating these genes reveals a slightly different phylogeny than the 16S one.
For comparison we compare this with the 40 genes deemed to be the most phylogenetically relevant at <a href="http://mltreemap.org/">mlTreeMap</a></p>

<figure style="text-align:center">
<img src="concribo.png" width="100%" />
<figcaption>A phylogenetic tree based on concatenated higly conserved ribosomal genes</figcaption>
</figure>

<figure style="text-align:center">
<img src="mltreemaptree.png" width="100%" />
<figcaption>A phylogenetic tree based on the concatenated genes from the mlTreeMap project</figcaption>
</figure>

<figure style="text-align:center">
<img src="concat_630.png" width="100%" />
<figcaption>A phylogenetic tree based on the concatenated genes from the core genome without the ribosomal genes</figcaption>
</figure>

<p>The other higly preserved cluster is a set of 14 NADH genes. It is an arc pointing east (and slightly south). These non ribosomal genes reveal a phylogeny more in line with what seen in the atlas. The <i>Thermus thermophilus</i> cluster merges with the cluster containing the <i>Thermus arciformis</i> and <i>Thermus kawarayensis</i></p>

<figure style="text-align:center">
<img src="concNADH.png" width="100%" />
<figcaption>A phylogenetic tree based on concatenated genes from a NADH cluster</figcaption>
</figure>
</section>

<section>
<b>Genome synteny in <i>Thermus</i> species</b>
<p>We draw a synteny gradient image of all completed <i>Thermus</i> species available at NCBI with the chromosome in <i>Thermus thermophilus HB8</i> as a reference.
The reference species genes are colored according to their location in the genome, giving a color gradient (from blue to green) from the origin of replication. 
This view gives a good indication of the gnome order in different strains and how segments are shuffled in different genomes.
The species order from the outermost circle is <i>Thermus thermophilus SG0</i>, <i>Thermus thermophilus HB8</i>, <i>Thermus thermophilus HB27</i>, 
<i>Thermus thermophilus JL-18</i>, <i>Thermus scotoductus SA-01</i>, <i>Thermus oshimai JL-2</i> and <i>Thermus sp. CCB</i>.
For the <i>Thermus thermophilus</i> strains we see obvious areas of chromosomal invertions. 
The area close to one fifth of the length of the chromosome in <i>Thermus thermophilus SG0</i> starting close to the origin of replication in the reference species
 (<i>HB8</i>) is a large inverted strand flanked on both ends by a chunk of tRNA-Gly, 5S and 23S ribosomal RNA. 
 As the synteny is quite well preserved in <i>Thermus thermophilus</i> and giving the long stretch of the repeat (3k bases), assemblers could have a hard time spoting the locations resulting in the wrong order of genes in the published genome.
Also noteworthy is the 'locational preference', the right and left side of the chart contains similar colors even though the species are different, 
indicating that the main mechanism of genome shuffling is invertions of not too large genome segments. We also see apparent jumps of segments. An out-of-place (gray) segment in the <i>Thermus oshimai JL-2</i> plasmid (compared to <i>Thermus thermophilus</i>) contains a cluster of thiamine/thiazole biosynthesis proteins, this cluster is not present in <i>Thermus kawarayensis</i> and <i>Thermus scotoductus 1572</i>.
Another segment which has different location than in the <i>Thermus thermophilus</i> strains contains menaquinone biosynthesis protein along with tryptophan and thymidilate synthase proteins.
A segment with response regulator containing CheY like receiver domain is also </p>
<figure style="text-align:center">
<img src="syntrainbow.png" width="100%" />
<figcaption>A synteny gradient rainbow</figcaption>
</figure>
<figure style="text-align:center">
<img src="isyntrainbow.png" width="100%" />
<figcaption>A inverted synteny gradient rainbow</figcaption>
</figure>
<p>We create a phylogenetic tree based on synteny. A distance matrix is created based on breakpoint count and neighbour joining algorithm used to construct the tree. 
Although perhaps not very reliable, this tree complies remarkably well to the ribosomal multiple gene tree.</p> 
<figure style="text-align:center">
<img src="syntree2.png" width="100%" />
<figcaption>A phylogenetic tree based genome synteny</figcaption>
</figure>
<p>Apart from those used to construct the phylogeny the genome synteny atlas reveals a few more areas with a preserved synteny in the <i>Thermus</i> species. 
Most notable are the UDP-N-acetylmuramate with a family of cell division proteins (arc facing west in the <i>Thermus thermophilus HB8</i> genome synteny atlas) and a cluster of Sulfur oxidation and cytochrome c proteins, 
although missing from the <i>Thermus scotoductus</i> typestrain (252_MATIS, DSM_8553) and <i>Thermus sp. RLM</i>. 
These strains are both lab strains which could explain their lack of Sulfur oxidizing ability.</p>
</section>
<section>
<b>Presence and absense of genes in <i>Thermus</i> lineages</b>
<p>Gene uptake and loss seems to follow phyloenetic paths.</p>
</section>
<section>
<b>Phage related genes and horizontal gene transfer in the <i>Thermus</i> species</b>
<p>The NCBI database contains five Thermus phages, the IN93, P23_45, P23_77, P74_26 and phyIS40. 
The genes from the P23_77 and the related IN93 are present in three icelandic strains, the <i>Thermus scotoductis 2127</i>, <i>Thermus oshimai</i> and <i>Thermus brockianus</i>. 
Also they are present in the <i>Thermus sp. RLM</i> and all the <i>Meithermus species</i> to some extent. This could indicate these species share a common niche.
</p>
<p>Different phylogeny is indicative of horizontaly transferred genes. 
A four gene cluster (energy production and conversion) containing a thiosulfate reduction, nrfC protein (iron sulfur) and anaeoribic dehydrogenase is highly conserved in some distantly related <i>Thermus</i> species while different in some more related species. 
Could be functional constraint or a sharing between species.</p>
<figure style="text-align:center">
<img src="nrfc.png" width="100%" />
<figcaption>A phylogenetic tree based on possibly horizontaly transferred genes. The tree is rooted on <i>Oceanithermus profundus</i></figcaption>
</figure>
<p>Some genes show an obvious phylogeographic pattern. 
For example a cluster of genes containing lysophospholipase, nmrA family protein and a UV damage endonuclease group species together according to location. The icelandic <i>Thermus scotoductus</i> with <i>Thermus islandicus</i> and The Yellowstone based <i>Thermus aquaticus</i>, <i>Thermus oshimai JL-2</i> and <i>Thermus thermophilus JL-18</i></p>
<figure style="text-align:center">
<img src="phylogeotree.png" width="100%" />
<figcaption>A phylogenetic tree with phylogeographic patterns</figcaption>
</figure>
<p>A cluster containing the Cytochrome P450, deoxyribodipyrimidine photolyase, Glycosyltransferases involved in cell wall synthesis and Phytoene synthase is conserved in the icelandic <i>Thermus scotoductus</i> and <i>Thermus islandicus</i>. The cluster is completely missing from other sequenced <i>Thermus scotoductus</i> and is not in the related icelandic <i>Thermus antranikianii</i></p>
<figure style="text-align:center">
<img src="phylogeotree2.png" width="100%" />
<figcaption>A phylogenetic tree with phylogeographic patterns</figcaption>
</figure>
</section>
<section>
<b>Metabolism in the <i>Thermus</i> species</b>
<p>The nitrate resipratory genes (alpha, beta and delta subunits NarG, NarH and NarI) are highly conserved but are missing from <i>Thermus aquaticus</i>, <i>Thermus islandicus</i>, <i>Thermus filiformis</i> and the Malaysian <i>Thermus sp. CCB</i>. 
Also lacking the Nitrate respiratory ability are the <i>Thermus thermophilus HB8, HB27</i> and <i>ATCC</i> and the two icelandic <i>Thermus scotoductus 346</i> and <i>2101</i>, hence this property seems quite sporatic</p>
<figure style="text-align:center">
<img src="nar-tree.png" width="100%" />
<figcaption>A tree showing the phylogeny of the Nar gene cluster in <i>Thermus</i> species</figcaption>
</figure>
<p>
Mannosylglycerate biosynthetic process is only found in the marine <i>Thermus thermophilus</i> although mannosyltransferase is also found in <i>Thermus islandicus</i> and <i>Thermus aquaticus</i>. 
Also found only in the <i>Thermus thermophilus</i> strains is a SCP-like extracellular protein, 
however lacking from these strains is the thymidine and deoxynucleoside kinases.
</p><p>
Histidine ammonia lyase is the <i>Meiothermus</i> species but only in <i>Thermus islandicus</i> and <i>Thermus kawaraysensis</i>.
The psrABC polysulfate reduction cluster found in <i>Thermus thermophilus HB27</i> and <i>Thermus scotoductus SA-01</i> is found in all 
<i>Thermus scotoductus</i> except for the typestrain (252, DSM8553). It is missing from the <i>Thermus islandicus</i>, <i>Thermus eggertsoni</i>, 
<i>Thermus aquaticus</i>, <i>Thermus filiformis</i> and all other <i>Thermus thermophilus</i>
</p><p>
<b>Transporters and Energy Metabolism</b>
</p><p>
The S-layer repressor found in <i>Thermus scotoductus SA-01</i> is present in all all the <i>Thermuse</i> strains except the <i>Thermus oshimai</i>, 
<i>Thermus brockianus</i>, <i>Thermus filiformis</i>, <i>Thermus aquaticus</i> and the <i>Thermus sp. CCB</i>, suggesting a sessile lifestyle.</p>
<p>
<b>Identification of competence proteins</b>
</p><p>
The competence proteins pilA, pilB, pilC, pilD, pilM, pilN, pilO and pilT were found in all the <i>Thermaceae</i> species.
The competence protein pilW was present in all <i>Thermus</i> strains but missing from other <i>Thermaceae</i> 
PilA were missing from the <i>Thermus scotoductus SA-01</i> but present in the same strain <i>4063</i> sequenced at MATIS.
comEA, comEC, cprA and comZ were all found in all <i>Thermaceae</i> strains.
</p>
<figure style="text-align:center">
<img src="p-pilus-tree.png" width="100%" />
<figcaption>The species containing a 6 gene cluster containing hypothetical genes and a P-Pilus assembly protein </figcaption>
</figure>
<figure style="text-align:center">
<img src="pilx_tree.png" width="100%" />
<figcaption>The species containing the ComZ protein</figcaption>
</figure>
<figure style="text-align:center">
<img src="cob.png" width="100%" />
<figcaption>A cluster containing cobalamin biosynthesis proteins and precorrin isomerase (CibC, CibD, CibG, CibH and CibX) present in the plasmids of the finished <i>Thermus</i> genomes. A high affinity nickel permease is found in the same species in addition to <i>Thermus islandicus</i>. CibB is found in all <i>Thermus</i> species.</figcaption>
</figure>
<figure style="text-align:center">
<img src="cytochrome_copper.png" width="100%" />
<figcaption>Cytochrome c oxidase subunit I and II are found in every species. A cluster containing another kind of Cytochrome C oxidase (I, II and III) along with Heme/Copper-type cytochrome/quinol oxidase is present in the <i>Thermus scotoductus</i> linage and <i>Thermus oshimai</i></figcaption>
</figure>
<p>myo-inosose-2 dehydratasex, myo-inositol catabolism (iolB, iolC, iolD, iolE and iolG) found in <i>Meiothermus ruber</i> strains is also found in the plasmid of <i>Thermus thermophilus SG0</i> and the <i>Thermus eggertsoni</i> and the icelandic <i>Thermus igniterrae</i>.
hydrogenase expression/formation protein and hydrogenase assembly chaperone (hypC, hypD and hypE) is present in the plasmid of <i>Thermus thermophilus SG0</i> and <i>Thermus oshimai DSM_12092</i>. A different kind is also in both <i>Thermus igniterrae</i>. The hydrogenase nickel insertion protein and hydrogenase accessory protein (hypA, hypB) are only missing from the <i>Thermus oshimai</i>
</p><p>
High affinity nickel transport protein is found in the icelandic <i>Thermus antranikianii</i>, the <i>Thermus islandicus</i> and two icelandic <i>Thermus scotoductus</i> the 2101 and 1572.
</p>
<figure style="text-align:center">
<img src="nife_hydrogenase.png" width="100%" />
<figcaption>Ni-Fe hydrogenase, Formate hydrogenlyase and hydrogenase 4 proteins form a cluser in the icelandic <i>Thermus species</i></figcaption>
</figure>
<p>
Formate hydrogenase (hyfB, hyfC, hyfE, hyfF and hyfG) are present in some icelandic <i>Thermus</i> strains.
</p>
</section>
</div>
</body>
</html>